### PR TITLE
Rework of Path of Gold and some Triforce Hunt changes

### PR DESCRIPTION
**Pull request by TestRunnerSRL**
Path of Gold is now replaced with "Opportunity" hints. Opportunity hints are weaker woth hints that are placed when there are optional goals and there are not enough woth locations. Optional goals refers to settings that set the required goals to be less than the available (such as medallion, stone, triforce hunt etc). These Opportunity hints are like woth, except that instead of being items required to beat the game, they are items required to complete all the reachable optional goals (such as collect all Triforces).

With this change, Triforce Hunt now only logically requires the minimum number of Triforces. This means that the Way of the Hero can now be generated for Triforce Hunt. The previous Path of Gold will now be Opportunity hints when there are not enough Way of the Hero. A consequence of this is that in Beatable-Only, some optional triforces may not be reachable.

Another change is that the number of triforces places is a setting along with the required pieces, instead of using the item distribution to determine the count. When the setting is randomized, it will default to 30 pieces, 20 required.

### CODE REVIEW

File: State.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def won(self):
 
 
     def won_triforce_hunt(self):
-        return self.has('Triforce Piece', self.world.triforce_count)
+        return self.has('Triforce Piece', self.world.triforce_goal)
```
Comment (Cuphat) [inline State.py:44]:
```suggestion
        return self.has('Triforce Piece', self.world.triforce_goal_per_world)
```

This should check only the Triforce Pieces in the world itself.

File: Hints.py [hunk 2]
```diff
@@ -322,6 +323,22 @@ def get_woth_hint(spoiler, world, checked):
...
+            location.name not in checked
+            and not (world.woth_dungeon >= world.hint_dist_user['dungeons_woth_limit'] and location.parent_region.dungeon)
+            and location.name not in world.hint_exclusions
+            and location.name not in world.hint_type_overrides['woth']
+            and location.item.name not in world.item_hint_type_overrides['woth'],
+            opportunity_locations))
+
+        # only select an from the opportunity locations if there are any
+        if opportunity_locations:
+            hint_suffix = "contains an opportunity."
```
Comment (fenhl) [inline Hints.py:339]:
```suggestion
            hint_suffix = "presents an opportunity."
```

Maybe this would sound a bit less dry?

**Comment by mracsys** [comment]
I've been giving this some thought and would like to propose some tweaks. I can put in the legwork on my branch if desired.

### Treat opportunities as their own hint type rather than a fallback for way of the hero
If fleshed out more, I can see opportunities completely replacing WOTH, or WOTH becoming a subset of opportunities with beating Ganon as one of the goals.

### Add goal "types" instead of global opportunities
Something like "X is on the path of the Forest" for items required to obtain the Forest Medallion. Hint distributions could draw from the list of types randomly or in a distributed manner. One guaranteed hint per path seems way too strong, but could be a replacement "very_strong" or new "beginner" distribution.

One downside to this approach is items could belong to multiple paths, but would only be hinted for one path with the current one-hint-per-location limit. Removing the limit may help balance the extra information provided, but could also provide too much info. I'm also unsure if the added goal info would be too strong, such as Slingshot on the path of the Spirit (vanilla Spirit medallion) with Mirror Shield on Gohma. With WOTH I think it's much more likely to be interpreted as needed for Deku completion if Deku is required. The counterargument could be for child projectiles for Spirit keys. Either way I think it would reward logic knowledge more than the current WOTH hints (could be good or bad). This may not be an issue in practice.

### Use modular goal list with plando support for custom goals
It could support hinting items for things like trials, or reaching specific high-sphere/gated locations in addition to bridge and BK conditions. Instead of hinting for obtaining the Forest Medallion in whatever temple, one could have a goal for completing the Forest Temple instead by reaching the Phantom Ganon boss heart item. I can also see this being a benefit for the bingo crowd to get in-logic hints for goals instead of the current logic-unaware named-item hints.

Rough structure would be:

* Goal
  * Path Name
  * Path Color
  * Get goals (list)
    * Item
    * Quantity (optional)
  * Reach goals (list)
    * Location
    * Age (optional)

**Comment by Zannick** [comment]
How are we on this?

**Comment by mracsys** [comment]
I didn't want to start work until it was discussed since this is TR's baby. I can get my described framework in place later this week 🙂.

**Comment by Zannick** [comment]
Sorry I didn't have a lot of time to really parse and discuss your proposal, @mracsys. I'm not certain yet what other people think of it, but if you're happy implementing it as an experiment, that's cool.

Would an item not required to beat the game be eligible for a hint like "path of the Forest"? I'm worried that having too many paths in a seed would be overwhelming for players to keep in their mind, and also whether it comes out as a recommendation to go to a particular place before another, even if neither of those places are required.

**Comment by mracsys** [comment]
No worries!

Non-required items could be hinted, yes. I hadn't thought that far. To fix that, the original design with opportunity hints as backup for WOTH can be adapted to the goal model by prioritizing goal hints that fulfill both a path and the global win condition. The difference is goal hint text is indistinguishable between WOTH and opportunity types. I think WOTH should still be a thing, but letting opportunity hints hijack them gives more info. It tells the player that the WOTH items are really important and that there aren't many since the hints ran out. Path hints give new info but each hint is about equal in value, and thus more resistant to seed variability. We'll see if that's true after I implement them.

**Comment by r0bd0g** [comment]
TR and I discussed a number of issues with this implementation and came up with a rather involved plan to fix them all, but I don't think any changes have actually been made since then. At this point I've forgotten what all the issues were, but I believe this current implementation has serious issues as far as actually working the way you might like it to.

**Comment by mracsys** [comment]
I'm not going to have time to work on my idea for the next couple weeks. I'll post changes here or in a separate PR when I am able.

**Comment by mracsys** [comment]
I put together initial goal support on my branch at https://github.com/mracsys/OoT-Randomizer/tree/goal-hints. Default goal requirements are always written to the spoiler regardless of if goal hints are on. The hint distribution in [goal-plando.txt](https://github.com/TestRunnerSRL/OoT-Randomizer/files/6017238/goal-plando.txt) will generate 1 goal hint and includes an example custom goal.

Location goals aren't fully implemented yet. Goal selection for hints is also completely random currently. I want to do some test races to see if this is enough or if weighted goal selection should be used.

When this is ready to PR should I make a separate one or try to merge it with this one?

There is overlap between the goal requirements and WOTH items for some win conditions, which makes sense. With dungeon rewards, later sphere dungeons also contain the same low sphere items. Looking at spoilers is difficult to get a feel for how they would play, but there is opportunity for more targeted hints with variable win conditions as seen in [OoT_6BFE8_PW7KMYZU42_Spoiler.txt](https://github.com/TestRunnerSRL/OoT-Randomizer/files/6017245/OoT_6BFE8_PW7KMYZU42_Spoiler.txt). The goals add quite a few more hintable locations that WOTH misses here (both hookshots, all 3 strength upgrades, iron boots, dins fire, mirror shield, hammer).

**Comment by r0bd0g** [comment]
I found a summary of the changes I felt were needed, added some clarifications:

1. In TP Hunt, only locations that are hard locking TPs should get opportunity hints. (This means that items that would be opportunity merely to open the bridge would not be considered -- if there's a TP locked by Ganon's it would only consider items that are WotH to open the rainbow bridge as an opportunity.)
2. LACS opportunities need to be processed separately from bridge opportunities. (This is mainly due to a conflict with point 3.)
3. Bridge opportunities should test reachability on the prizes that unlock it by assuming the bridge is not repaired. (To prevent locations from being opportunities when you must spawn the bridge to reach them in the first place.)

**Comment by fenhl** [comment]
> Related to point 1, there's the 'all goals reachable' setting that includes access to Ganon's even on triforce hunt (actually, just the rewards that open the bridge; access to Ganon's might not actually be possible). There's not really a great place to suggest that be changed but I'll voice my objection to the current implementation here, I guess, since this is kind of related to that.

I think the AGR behavior should match that of opportunity hints as closely as possible, so I'll make another PR to change it if the hint behavior is changed.

**Comment by fenhl** [comment]
I believe this can be closed now, since all changes proposed here were merged separately:

* Opportunity hints were further developed into goal hints in #1361.
* #1419 disabled the special requirement for all Triforce pieces to be reachable even if All Locations Reachable is off. A setting to require all goal items to be reachable, similar to the old behavior but also applying to other types of goals, was added in #1198.
* The number of pieces placed was made into its own setting in #1403.

**Pull request closed without merging**